docs/architecture/adr-018-ABCI-Validators.md

```diff
@@ -25,6 +25,7 @@ Validator (with just the address) and a ValidatorUpdate (with the pubkey):
 ```
 message Validator {
     bytes address
```

Comment:
I don't understand what we gain by referring to validators by address. The downstream application has to maintain a map from address -> pubkey, which presumably requires another hashmap. (Or maybe a treemap, thats probably faster at the input size)

So we now are doing 2 hashes (one cryptographic, one quick) instead of just sending over the pubkey bytes. I don't understand what efficiency gain we're getting over ABCI with using the address here. It only shaves off 12 bytes due to tmhash (maybe less after #1990). Using address here saves 12 bytes of inter process communication (likely will be less soon), and requires an extra hash to be called on the receivers end, and more mental overhead for the address -> pubkey relation. 

I think everything will be simpler if we just used pubkey here. Am I missing something? If we wanted to change the pubkey at an address, we already have to set the current power to 0 and use another pubkey with the current implementation.

Reply:
> Am I missing something?

With quantum pubkeys, addresses save kilobytes. 

Maybe it's a premature optimization, but the point is so ABCI devs are forced to be aware of addresses and the difference between them and pubkeys, and basically encouraged to use addresses as the native index for their pubkeys.

Reply:
Mentioned this there as well:
> Address probably makes a lot of sense over ABCI* when we go to post quantum public keys, but thats very #postlaunch and probs not worth doing right now.

\* previously wrote IBC, meant ABCI

I think we're over 1-2 years out before it makes sense to start supporting the switch to quantum pubkeys on the validator set. Right now we're looking at switching to BLS for at least awhile. Then (I imagine) we are going to want to switch to bilinear map solution on lattices. BLS12-381 pubkeys are 94 bytes. I'm not convinced that double hashing the 94 byte case will be more efficient than just sending the extra data over the wire. (One extra sha2 block-hash vs sending 74 extra bytes over the wire)

> Maybe it's a premature optimization, but the point is so ABCI devs are forced to be aware of addresses and the difference between them and pubkeys, and basically encouraged to use addresses as the native index for their pubkeys.

This definitely feels like a premature optimization to me. I think we're quite far from post-quantum. If we decided that we wanted to do this for BLS, then this would make sense to me as something to do prelaunch. 

If we want to still go ahead with this change, can we note this somewhere? (The reasoning is sound, just depends on if we want to make this change now, or when we approach the quantum era)

Reply:
Ref https://github.com/cosmos/cosmos-sdk/pull/1927#issuecomment-411362312

I'm not convinced the downstream application will need to maintain a hashmap - in which situations would the ABCI app absolutely need to know the pubkey? Evidence handled in the state machine instead of in Tendermint?

Reply:
Could we send the validator address instead of the validator public key back to Tendermint to update the validator set instead?

Reply:
I think a better solution would just be only use the pubkey. We're not getting any computational benefit right now with the address + mental overhead of whether to use address or pubkey.

If we made updates use the address we would have to have a separate mechanism to make tendermint learn the pubkey -- even less efficient and more complex.

Reply:
> Could we send the validator address instead of the validator public key back to Tendermint to update the validator set instead?

This would require the first vote from a new validator to then include the pubkey. We had some discussion about it before: https://github.com/tendermint/tendermint/issues/1625

> I think a better solution would just be only use the pubkey

I like to agree, but Jae's concern was that since we're going to need to switch to addresses in the the future to accomodate quantum pubkeys (realistically it could be in high demand next year, at least to start experimenting with on systems like Tendermint), it will require significant restructuring of apps that don't already have some kind of map from addresses to pubkeys.

Reply:
Lets benchmark how much time reading 3kb takes. I'm still not fully convinced that the overhead of hashing 3kb will outweigh the time to read it / ABCI decode it.

Reply:
Perhaps the solution is to have a transitionary period right now where we pass both the address and pubkey on BeginBlock, and expect to stop passing the pubkey at some point in the future, say in 6 months or something. Of course that will just punt the need to refactor SDK apps until then, which may make it more difficult at that time. 

We could reconsider the "apps only need to know about addresses" thing, but then we have to accomodate the "pubkey on the first vote" tendermint side, and I do agree with jae that we shoudn't completely isolate apps from the pubekys because they're at the essence of the signature mechanism - address is just a reference

Reply:
Just benchmarked it, it definitely seems like addresses are the wrong move even _after_ quantum pubkeys. 
```
BenchmarkSHA256_3kb-8    	  200000	      8168 ns/op
BenchmarkMake_3kb-8      	30000000	        48.3 ns/op
```
If we were to do the address method, thats 8000+ ns PLUS memory of the hashmap PLUS time for AES hash. (Rough AES hash time estimate based off of https://github.com/tendermint/tendermint/pull/2115#discussion_r206631957 is 210 ns) The time to create 3kilo bytes is 48ns on my machine. If we're allocating memory over 150 times for a pubkey, were doing something _extremely_ wrong. I think we should just always use the pubkey, its not like its going over the internet here.

Reply:
Actually another thing came up in Chris reviewing my PR. The pubkey is actually going over the internet an extra time in syncing when we maintain the addr pubkey map, as we have to place this map in state. So we should either reconsider "apps only need to know about addresses", or only use pubkeys. (I think the latter is more sensible, the extra 3kb memcopy + a couple of allocations nowhere near makes up for hash the cost.

Reply:
Jae made the point that when we go to post-quantum signature schemes, IAVL traversal of keys in staking will take longer. 

I think we should either go with your suggestion @ebuchman (to use both) or just use pubkeys at launch. Makes the code simpler on the SDK side for launch, and we need to fully weigh hash time vs IAVL traversal time if its for efficiency reasons. (Shouldn't make plans for optimization without benchmarks!)

Reply:
> Just benchmarked it,

Not sure about the SHA vs Make benchmark. Seems the real benchmark would be "SHA and marshal, send, unmarshal Validators *without* Pubkey over ABCI socket" vs. "no SHA and marshal, send, unmarshal Validators *with* Pubkey over ABCI socket"

> The pubkey is actually going over the internet an extra time in syncing when we maintain the addr pubkey map, as we have to place this map in state

What does keeping the map in state have to do with going over the internet? It's all computed from data we already have.

Reply:
We don't have to compute the address each time we send the address over.  If we optimize Tendermint it should be something that is amortized over time.  Which leads to pure ABCI bandwidth.

We should continue with addresses and either support it with plans to deprecate it, or only require addresses.

Reply:
> We should continue with addresses and either support it with plans to deprecate it, or only require addresses.

What do you mean by plans to deprecate it here?